Marlin/src/HAL/STM32F1/fast_pwm.cpp

```diff
@@ -0,0 +1,111 @@
+/**
+ * Marlin 3D Printer Firmware
+ * Copyright (c) 2020 MarlinFirmware [https://github.com/MarlinFirmware/Marlin]
+ *
+ * Based on Sprinter and grbl.
+ * Copyright (c) 2011 Camiel Gubbels / Erik van der Zalm
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+#ifdef __STM32F1__
+
+#include "../../inc/MarlinConfigPre.h"
+
+#if NEEDS_HARDWARE_PWM // Specific meta-flag for features that mandate PWM
+
+#include <pwm.h>
+#include "HAL.h"
+#include "timers.h"
+
+void set_pwm_frequency(const pin_t pin, int f_desired) {
+    if(!PWM_PIN(pin)) return;   //don't proceed if no hardware timer
+
+    timer_dev* timer = PIN_MAP[pin].timer_device;
+    uint8_t chanel = PIN_MAP[pin].timer_channel;
+
+    //protect used timers
+    if(timer == get_timer_dev(STEP_TIMER_NUM)) return;
+    if(timer == get_timer_dev(PULSE_TIMER_NUM)) return;
+    if(timer == get_timer_dev(TEMP_TIMER_NUM)) return;
```

Comment:
PULSE timer doesn't really exist, it is just the STEP_TIMER on this platform.

There are more timers than this which could be interfered with:
1. Servo Timer
2. Tone Timer
3. Software Serial Timer
4. Timers connected to SPI pins

The SPI bit might seem weird, but I believe on STM32F1 chips pins are not actually disconnected from the timers even if you are using their other functions. This means that running the timer connected to SPI interferes with SPI, even if you are running it for other purposes.

Detecting all the possible timer conflicts is _hard_, which is why there have been so many mistakes made in this area.
Using your feature won't be any less safe than "normal" PWM on these pins, so perhaps is little to be gained from these partial checks.

Reply:
> 1. Servo Timer
> 2. Tone Timer
> 3. Software Serial Timer
> 4. Timers connected to SPI pins

Am I right, that all timers, allocated by Marlin are listed in timers.h? Or there any implicit usage? (except libraries, but it will be good to know what timers exactly they allocates)

Reply:
Some are implicit. I think that dealing with timers is a little more mature in  HAL/STM32. Maple (HAL/STM32F1) doesn't always give a good way to override the default behavior.

Tone is the hardest to deal with in Maple. It default to 4 or 8 (depending on the chip), but will use the timer connected to a pin if available. This means it is entirely up to the board layout whether enabling SPEAKER will interfere with some other timer.